File: IDWiseDynamicJourney/ios/IDWiseModule.swift
//
//  IDWiseSDKBridge.swift
//  idwiseReactNativeSample
//
//  Created by Hafiz Ahsan on 14/04/2022.
//

import Foundation
import IDWiseSDK
import React

@objc (IDWiseModule)
class IDWiseModule: RCTEventEmitter  {
  
  var journeyID = ""
  
  public static var emitter: RCTEventEmitter!
  
  open override func supportedEvents() -> [String] {
    ["onJourneyStarted","onJourneyResumed", "onJourneyCancelled" ,"onError", "onJourneyFinished", "onJourneySummary", "onStepCaptured", "onStepResult", "onStepCancelled", "onStepSkipped"]
   }
  
  override init() {
    super.init()
    IDWiseModule.emitter = self
  }
  @objc func initialize(_ clientKey: String,_ theme: String) {
    var sdkTheme: IDWiseSDKTheme = .systemDefault
    if theme == "LIGHT" {
      sdkTheme = .light
    } else if theme == "DARK" {
      sdkTheme = .dark
    }
    IDWise.initialize(clientKey: clientKey,theme: sdkTheme) { err in
          if let error = err {
            IDWiseModule.emitter.sendEvent(withName: "onError", body: ["errorCode": error.code, "errorMessage": error.message] as [String : Any])
          }
        }
    
  }
  
  @objc func startDynamicJourney(_ journeyDefinitionId:String,
                          _ referenceNo: String,
                          _ locale: String) {
    if referenceNo.isEmpty {
      DispatchQueue.main.async {
        IDWise.startDynamicJourney(journeyDefinitionId: journeyDefinitionId,
                                   locale: locale,
                                   journeyDelegate: self,
                                   stepDelegate: self)
      }
    } else {
      DispatchQueue.main.async {
        IDWise.startDynamicJourney(journeyDefinitionId: journeyDefinitionId,
                                   referenceNumber: referenceNo,
                                   locale: locale,
                                   journeyDelegate: self,
                                   stepDelegate: self)
      }
    }
  }
  
  @objc func resumeDynamicJourney(_ journeyDefinitionId:String,
                                  _ journeyId: String,
                                  _ locale: String) {
    DispatchQueue.main.async {
      IDWise.resumeDynamicJourney(journeyDefinitionId: journeyDefinitionId, journeyId: journeyId, locale: locale, journeyDelegate: self, stepDelegate: self)
    }
  }
  
  @objc func startStep(_ stepId: String) {
    DispatchQueue.main.async {
      IDWise.startStep(stepId: stepId)
    }
  }
  
  @objc func startStepFromFileUpload(_ stepId: String,
                                     _ data: Data) {
    DispatchQueue.main.async {
      IDWise.startStepFromFileUpload(stepId: stepId, data: data)
    }
  }
  
  @objc func unloadSDK() {
    DispatchQueue.main.async {
      IDWise.unloadSDK()
    }
  }
  
  @objc func getJourneySummary() {
    
    IDWise.getJourneySummary { journeySummary,error in
      if let summary = journeySummary {
        
        // encoding JSON response so react-native can consume it correctly
        var journeyStepSummaries = ""
        var journeyResult = ""
        
        do {
          let encodedStepSummaries = try JSONEncoder().encode(summary.stepSummaries)
          if let jsonString = String(data: encodedStepSummaries,
                                     encoding: .utf8) {
            journeyStepSummaries = jsonString
          }
        
          let encodedJourneyResult = try JSONEncoder().encode(summary.journeyResult)
          if let jsonString = String(data: encodedJourneyResult,
                                     encoding: .utf8) {
            journeyResult = jsonString
          }
          
        } catch {}

        IDWiseModule.emitter.sendEvent(withName: "onJourneySummary", body: ["journeyId": summary.journeyId,
                                                                            "journeyStepSummaries": journeyStepSummaries,
                                                                            "journeyResult": journeyResult,
                                                                            "journeyIsComplete": summary.isCompleted] as [String : Any])
      } else {
        if let err = error {
          IDWiseModule.emitter.sendEvent(withName: "onError", body: ["errorCode": err.code, "errorMessage": err.message] as [String : Any])
        }
      }
      
    }
  }

}

extension IDWiseModule: IDWiseSDKJourneyDelegate {
  @objc func JourneyStarted(journeyID: String) {
    self.journeyID = journeyID
    IDWiseModule.emitter.sendEvent(withName: "onJourneyStarted", body: ["journeyId": journeyID])
    print("Journey started with journey Id : \(journeyID)")
    getJourneySummary()
  }
  
  func onJourneyResumed(journeyID: String) {
    IDWiseModule.emitter.sendEvent(withName: "onJourneyResumed", body: ["journeyId": journeyID])
    print("Journey resumed with journey Id : \(journeyID)")

  }
  
  func JourneyFinished() {
    IDWiseModule.emitter.sendEvent(withName: "onJourneyFinished", body: ["journeyId": self.journeyID])
    print("Journey Finished")
  }
  
  func JourneyCancelled() {
    print("Journey Cancelled")
    IDWiseModule.emitter.sendEvent(withName: "onJourneyCancelled", body: ["journeyId": self.journeyID])

  }
  
  func onError(error: IDWiseSDKError) {
    IDWiseModule.emitter.sendEvent(withName: "onError", body: ["errorCode": error.code, "errorMessage": error.message] as [String : Any])  }
}

extension IDWiseModule: IDWiseSDKStepDelegate {
  func onStepCancelled(stepId: String) {
    IDWiseModule.emitter.sendEvent(withName: "onStepCancelled", body: ["stepId": stepId])
  }
  
  func onStepSkipped(stepId: String) {
    IDWiseModule.emitter.sendEvent(withName: "onStepSkipped", body: ["stepId": stepId])
  }
  
  func onStepCaptured(stepId: Int, capturedImage: UIImage?) {
    IDWiseModule.emitter.sendEvent(withName: "onStepCaptured", body: ["stepId": stepId])
  }
  
  func onStepResult(stepId: Int, stepResult: StepResult?) {
    
    // encoding JSON response so react-native can consume it correctly

    var stepResultObj = ""
    do {
      let encodedStepResult = try JSONEncoder().encode(stepResult)
      if let jsonString = String(data: encodedStepResult,
                                 encoding: .utf8) {
        stepResultObj = jsonString
      }
    } catch {}
    
    IDWiseModule.emitter.sendEvent(withName: "onStepResult", body: ["stepId": stepId,"stepResult": stepResultObj] as [String : Any])
  }
  
  func onStepConfirmed(stepId: String) {
    
  }
  
  
}

File: IDWiseDynamicJourney/android/app/src/main/java/com/idwisereactnativesample/IDWiseModule.java
package com.idwisereactnativesample;

import android.annotation.SuppressLint;
import android.graphics.Bitmap;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.UiThreadUtil;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.google.gson.Gson;
import com.idwise.sdk.IDWise;
import com.idwise.sdk.IDWiseSDKCallback;
import com.idwise.sdk.IDWiseSDKStepCallback;
import com.idwise.sdk.data.models.IDWiseSDKError;
import com.idwise.sdk.data.models.IDWiseSDKTheme;
import com.idwise.sdk.data.models.JourneyInfo;
import com.idwise.sdk.data.models.JourneySummary;
import com.idwise.sdk.data.models.StepResult;

import kotlin.Unit;
import kotlin.jvm.functions.Function1;
import kotlin.jvm.functions.Function2;

public class IDWiseModule extends ReactContextBaseJavaModule {

    ReactApplicationContext context;
    IDWiseSDKCallback journeyCallback;
    IDWiseSDKStepCallback stepCallback;


    IDWiseModule(ReactApplicationContext context) {
        super(context);
        this.context = context;
    }

    @NonNull
    @Override
    public String getName() {
        return "IDWiseModule";
    }


    /**
     * this function is used to send back the events related to Verification
     * Journey, to the JavaScript interface
     */

    private void sendEvent(String eventName, @Nullable WritableMap params) {
        context.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, params);
    }

    @ReactMethod
    public void initialize(String clientKey, String theme) {
        IDWiseSDKTheme idwiseTheme;
        try {
            idwiseTheme = IDWiseSDKTheme.valueOf(theme);
        } catch (Exception e) {
            idwiseTheme = IDWiseSDKTheme.SYSTEM_DEFAULT;
        }

        Function1<? super IDWiseSDKError, Unit> errorCallback = (Function1<IDWiseSDKError, Unit>) error -> {

            WritableMap params = Arguments.createMap();
            params.putInt("errorCode", error.getErrorCode());
            params.putString("errorMessage", error.getMessage());
            sendEvent("onError", params);

            return null;
        };

        IDWise.INSTANCE.initialize(clientKey, idwiseTheme, errorCallback);
    }

    @ReactMethod
    public void startDynamicJourney(String journeyTemplateId,
                                    String referenceNo,
                                    String locale) {
        setJourneyCallback();
        setStepCallback();

        IDWise.INSTANCE.startDynamicJourney(
                getCurrentActivity(),
                journeyTemplateId,
                referenceNo,
                locale,
                journeyCallback,
                stepCallback
        );
    }

    @ReactMethod
    public void resumeDynamicJourney(String journeyTemplateId,
                                     String referenceNo,
                                     String locale) {
        setJourneyCallback();
        setStepCallback();

        IDWise.INSTANCE.resumeDynamicJourney(
                getCurrentActivity(),
                journeyTemplateId,
                referenceNo,
                locale,
                journeyCallback,
                stepCallback
        );
    }

    @ReactMethod
    public void startStep(String stepId) {
        IDWise.INSTANCE.startStep(getCurrentActivity(), stepId);
    }

    @ReactMethod
    public void startStepFromFileUpload(String stepId, byte[] data) {
        IDWise.INSTANCE.startStepFromFileUpload(getCurrentActivity(), stepId, data);
    }

    @ReactMethod
    public void getJourneySummary(String journeyId) {
        Function2<? super JourneySummary, ? super IDWiseSDKError, Unit> summaryCallback = (Function2<JourneySummary, IDWiseSDKError, Unit>) (summary, error) -> {

            WritableMap params = Arguments.createMap();

            if (summary != null) {
                Gson gson = new Gson();
                params.putString("journeyId", summary.getJourneyId());
                params.putString("journeyStepSummaries", gson.toJson(summary.getStepSummaries()));
                params.putString("journeyDefinition", gson.toJson(summary.getJourneyDefinition()));
                params.putString("journeyResult", gson.toJson(summary.getJourneyResult()));
                params.putBoolean("journeyIsComplete", summary.isCompleted());
                sendEvent("onJourneySummary", params);
            }

            if (error != null) {
                params.putInt("errorCode", error.getErrorCode());
                params.putString("errorMessage", error.getMessage());
                sendEvent("onError", params);
            }

            return null;
        };

        IDWise.INSTANCE.getJourneySummary(journeyId, summaryCallback);
    }

    @ReactMethod
    public void unloadSDK() {
        UiThreadUtil.runOnUiThread(IDWise.INSTANCE::unloadSDK);
    }


    @ReactMethod
    public void addListener(String eventName) {

    }

    @ReactMethod
    public void removeListeners(Integer count) {

    }

    private void setJourneyCallback() {
        journeyCallback = new IDWiseSDKCallback() {
            @Override
            public void onJourneyStarted(@NonNull JourneyInfo journeyInfo) {
                WritableMap params = Arguments.createMap();
                params.putString("journeyId", journeyInfo.getJourneyId());
                sendEvent("onJourneyStarted", params);
            }

            @Override
            public void onJourneyResumed(@NonNull JourneyInfo journeyInfo) {
                WritableMap params = Arguments.createMap();
                params.putString("journeyId", journeyInfo.getJourneyId());
                sendEvent("onJourneyResumed", params);
            }

            @Override
            public void onJourneyCompleted(@NonNull JourneyInfo journeyInfo, boolean isCompleted) {
                WritableMap params = Arguments.createMap();
                params.putString("journeyId", journeyInfo.getJourneyId());
                sendEvent("onJourneyFinished", params);
            }

            @Override
            public void onJourneyCancelled(@Nullable JourneyInfo journeyInfo) {
                WritableMap params = Arguments.createMap();
                params.putString("journeyId", journeyInfo == null ? null : journeyInfo.getJourneyId());
                sendEvent("onJourneyCancelled", params);
            }

            @Override
            public void onError(@NonNull IDWiseSDKError idWiseSDKError) {

                WritableMap params = Arguments.createMap();
                params.putInt("errorCode", idWiseSDKError.getErrorCode());
                params.putString("errorMessage", idWiseSDKError.getMessage());
                sendEvent("onError", params);

            }
        };
    }

    private void setStepCallback() {
        stepCallback = new IDWiseSDKStepCallback() {
            @SuppressLint("CheckResult")
            @Override
            public void onStepCaptured(@NonNull String s, @Nullable Bitmap bitmap, @Nullable Bitmap croppedBitmap) {
                WritableMap params = Arguments.createMap();
                params.putString("stepId", s);
                sendEvent("onStepCaptured", params);
            }

            @Override
            public void onStepResult(@NonNull String s, @Nullable StepResult stepResult) {
                WritableMap params = Arguments.createMap();
                params.putString("stepId", s);
                params.putString("stepResult", new Gson().toJson(stepResult));
                sendEvent("onStepResult", params);
            }

            @Override
            public void onStepConfirmed(@NonNull String s) {
                /**
                 * No need to implement this as for latest version it's not being required.
                 */
            }
        };
    }
}
